Adding simple API to get status/terminate/raise-event URLs

gled4er opened the pull request:
Hello @cgillum ,

May I ask you to review the direction I took for implementing issue [#155](https://github.com/Azure/azure-functions-durable-extension/issues/155#issuecomment-391624661)?

I tried to reduce the duplication as much as possible.

But let me know if you prefer to have dedicated method for the new API.

I also did not add any new code to the `DurableActivityContext` because we don't have dependency on `DurableTaskExtension` where currently the main method I use is implemented. Also I think that it is important to have the method available in `DurableOrchestrationContext` and if we need it in `DurableActivityContext` we can easily pass it.  

Please let me know if this is acceptable approach or it is recommended to add logic to `DurableActivityContext`  as well.

Thank you!

cgillum commented on test/HttpApiHandlerTests.cs:
```diff
@@ -73,6 +73,78 @@ private void CreateCheckStatusResponse_Throws_Exception_When_NotificationUrl_Mis
+            Assert.NotNull(checkStatus);
+            Assert.Equal(checkStatus.Id, TestConstants.InstanceId);
+            Assert.Equal(
+                "http://localhost:7071/admin/extensions/DurableTaskExtension/instances/7b59154ae666471993659902ed0ba742?taskHub=DurableFunctionsHub&connection=Storage&code=mykey",
```
Is the `http://localhost:7071/admin/extensions/DurableTaskExtension` based on `TestConstants.NotificationUrl`? If so, this might be a good place to reduce string duplication. I mention this mainly because Functions v2 will have a different URL prefix, and it might be nice to be able to update the constant value in one place and have these tests continue to work.

gled4er replied:
Thank you! I will change it.

cgillum commented on test/HttpApiHandlerTests.cs:
```diff
@@ -73,6 +73,78 @@ private void CreateCheckStatusResponse_Throws_Exception_When_NotificationUrl_Mis
+                "http://localhost:7071/admin/extensions/DurableTaskExtension/instances/7b59154ae666471993659902ed0ba742/terminate?reason={text}&taskHub=SampleHubVS&connection=TestConnection&code=mykey",
+                checkStatus.TerminatePostUri);
+        }
+
```
It's great to have these unit tests to help ensure we don't accidentally break the URL formats - thanks for adding them!

gled4er replied:
Sure. Tests are super useful to validate all expected behaviors.

gled4er commented:
Hello @cgillum,

Thank you very much for the comments.

I will work on fixing the issues next.

Thank you!

gled4er commented on src/WebJobs.Extensions.DurableTask/CheckStatus.cs:
```diff
@@ -0,0 +1,37 @@
+    /// <summary>
+    /// Data structure containing status, terminate and send external event HTTP endpoints
+    /// </summary>
+    public class CheckStatus
```
Great! I will change it! Thank you!

gled4er commented on src/WebJobs.Extensions.DurableTask/DurableOrchestrationContextBase.cs:
```diff
@@ -390,10 +391,17 @@ public virtual Task CreateTimer(DateTime fireAt, CancellationToken cancelToken)
         public abstract void SetCustomStatus(object customStatusObject);
+
+        /// <summary>
+        /// Creates a <see cref="CheckStatus"/> object that is useful for checking the status of the specified instance.
```
Thank you! I will update them.

gled4er commented:
Hello @cgillum,

May I ask you to check the current state of the pull request?

I worked on all issues you raised.

Please let me know what should be the next items for fixing.

Thank you!

gled4er commented on test/TestHelpers.cs:
```diff
@@ -236,6 +244,61 @@ private static List<string> GetLogs_HelloWorldOrchestration_Inline(string instan
             return list;
         }
 
+        private static List<string> GetLogs_Orchestrator_Retreives_HttpManagementPayload(string instanceId, string[] functionNames, bool extendedSessions)
```
Sure. I will be careful not to add trace validation. Do you prefer to remove it from the current PR?

gled4er commented on test/TestOrchestrations.cs:
```diff
@@ -19,6 +19,12 @@ public static string SayHelloInline([OrchestrationTrigger] DurableOrchestrationC
 
+        public static HttpManagementPayload GetHttpManagementPayload([OrchestrationTrigger] DurableOrchestrationContext ctx)
+        {
+            HttpManagementPayload httpManagementPayload = ctx.CreateHttpManagementPayload(ctx.InstanceId);
```
I explained a bit in the comment below. My thinking is that passing the `HttpManagementPayload`  information from the client to the orchestrator and the activity is currently not very useful. I think the most useful use cases is getting `HttpManagementPayload` from the orchestrator and then providing it to any activity that needs it. Maybe I missing part of your idea for design? 

The problem I see with having access to `HttpManagementPayload` only in the client is that first we need to start the orchestration and then to pass `HttpManagementPayload`. I see this working well only when the orchestrator is waiting for external event which make the whole flow a bit heavy in my opinion.

Please feel free to correct my understanding.

cgillum replied:
I agree trying to pass this data from the client to the orchestrator to the activity is not very useful. However, my idea was a little different, so I apologize for not being clear. My thought was to simply have an activity function *in the middle of an orchestration* which uses the `[OrchestrationClient]` binding to get an instance of `HttpManagementPayload`. For example:

```csharp
public static void SendInstanceInfo(
    [ActivityTrigger] DurableActivityContext ctx,
    [OrchestrationClient] DurableOrchestrationClient client)
{
    var payload = client.CreateHttpManagementPayload(ctx.InstanceId);
    // send the payload to some external destination
}
```
Let me know what you think.

gled4er replied:
Hello @cgillum,

This is great! Thank you very much for the sample! This made it clear for me. I will remove the code related to `DurableOrchestrationContext`. 

I agree with you that we should not include API methods to `DurableOrchestrationContext` if we don't need them. We need to access `HttpManagementPayload` in the activity functions. So support in `DurableOrchestrationClient` will be enough!

Thank you!

gled4er commented on test/DurableTaskEndToEndTests.cs:
```diff
@@ -1580,6 +1580,162 @@ private static void CompareTwoComplexTypeObjects(ComplexType firstObject, Comple
+        /// </summary>
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
```
This is great clarification! Do you think it make sense to start maintaining a wiki with all similar tips and guidance in order to reduce your review time?

cgillum commented:
Sorry, I think I didn't not communicate clearly enough. I mentioned this comment directly in the code, but I'll paste it here as well so it doesn't get lost:

I agree trying to pass this data from the client to the orchestrator to the activity is not very useful. However, my idea was a little different, so I apologize for not being clear. My thought was to simply have an activity function in the middle of an orchestration which uses the `[OrchestrationClient]` binding to get an instance of `HttpManagementPayload`. For example:

```csharp
public static void SendInstanceInfo(
    [ActivityTrigger] DurableActivityContext ctx,
    [OrchestrationClient] DurableOrchestrationClient client)
{
    var payload = client.CreateHttpManagementPayload(ctx.InstanceId);
    // send the payload to some external destination
}
```
Let me know what you think.

gled4er commented:
Hello @cgillum,

I removed the code relegated to `DurableOrchestrationContext` and I modified one of the existing integration tests to follow the sample you gave me.

Please when you have time take a look at the current state of the code and let me know what should be fixed next.

Thank you!

cgillum commented on test/TestActivities.cs:
```diff
@@ -86,9 +86,11 @@ public static double BindToDouble([ActivityTrigger] double value)
         }
 
-        public static HttpManagementPayload GetAndPassHttpManagementPayload([ActivityTrigger] DurableActivityContext ctx)
+        public static HttpManagementPayload GetAndPassHttpManagementPayload(
```
Let's rename this to `GetAndReturnHttpManagementPayload`.

gled4er replied:
Thank you! I will change it.

cgillum commented on test/TestHelpers.cs:
```diff
@@ -275,7 +257,7 @@ private static List<string> GetLogs_Activity_Gets_HttpManagementPayload(string i
                     $"{instanceId}: Function '{functionNames[1]} ({FunctionType.Activity})' scheduled. Reason: ReturnHttpManagementPayload. IsReplay: False. State: Scheduled. HubName: {taskHubName}. AppName: . SlotName: . ExtensionVersion: 1.4.1.0. SequenceNumber: 2.",
                     $"{instanceId}: Function '{functionNames[0]} ({FunctionType.Orchestrator})' awaited. IsReplay: False. State: Awaited. HubName: {taskHubName}. AppName: . SlotName: . ExtensionVersion: 1.4.1.0. SequenceNumber: 3.",
-                    $"{instanceId}: Function '{functionNames[1]} ({FunctionType.Activity})' started. IsReplay: False. Input: [{{\"$type\":\"Microsoft.Azure.WebJobs.Extensions.DurableTask.HttpManagementPayload, Microsoft.Azure.WebJobs.Extensions.DurableTask\",\"id\":\"{instanceId}\",\"statusQueryGetUri\":\"{TestConstants.NotificationUrlBase}/instances/{instanceId}?taskHub={taskHubName}&connection= ...
+                    $"{instanceId}: Function '{functionNames[1]} ({FunctionType.Activity})' started. IsReplay: False. Input: [null]. State: Started. HubName: {taskHubName}. AppName: . SlotName: . ExtensionVersion: 1.4.1.0. SequenceNumber: 4.",
```
Let's remove all the text at `HubName` and beyond. We don't want to test for this because that will cause test failures when, for example, we update the extension version. The test code does a prefix comparison, so the full trace doesn't need to match.

cgillum commented on test/TestHelpers.cs:
```diff
@@ -236,6 +241,46 @@ private static List<string> GetLogs_HelloWorldOrchestration_Inline(string instan
             return list;
         }
 
+        private static List<string> GetLogs_Activity_Gets_HttpManagementPayload(string instanceId, string[] functionNames, bool extendedSessions)
```
Can we actually remove these log checks too? I think the other components of your end-to-end test are already sufficient.

gled4er commented:
Hello @cgillum,

Thank you for the great feedback.

I will work on fixing the issues.

Thank you!

gled4er commented:
Hello @cgillum,

I modified the code according your recommendations.

Please check once more when you have time and let me know what should be fixed next.

Thank you!

gled4er commented:
Hello @cgillum,

Thank you very much for the support and especially for the sample that clarify the design for me!

Thank you!

gled4er commented:
Hello @cgillum,

Shall we update the documentation with this scenario? If so where do you want to add the new feature?

Thank you!

cgillum commented:
Yes, we should update the documentation.  I think the existing **Instance Management** topic is the right place for this. Thanks!

gled4er commented:
Hello @cgillum,

I will work on that.

Thank you!
